tinybytes/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+    /// `&[u8]` which is in fact a slice of the `Bytes`, i.e. a subset of it.
+    /// This function turns that `&[u8]` into another `Bytes`, as if one had
+    /// called `self.slice()` with the range that corresponds to `subset`.
+    ///
+    /// This operation is `O(1)`.
+    pub fn slice_ref(&self, subset: &[u8]) -> Option<Bytes> {
+        // An empty slice can be a subset of any slice.
+        if subset.is_empty() {
+            return Some(Bytes::empty());
+        }
+
+        let subset_start = subset.as_ptr() as usize;
+        let subset_end = subset_start + subset.len();
+        let self_start = self.ptr as usize;
+        let self_end = self_start + self.len;
+        if subset_start >= self_start && subset_end <= self_end {
+            Some(self.safe_slice_ref(subset_start - self_start, subset_end - self_start))
+        } else {
+            None
+        }
+    }
+
+    // private
+
+    fn from_underlying(value: impl UnderlyingBytes) -> Self {
+        let slice: &[u8] = value.as_ref();
+        Self {
+            ptr: slice.as_ptr(),
+            len: slice.len(),
+            bytes: Some(Arc::new(value)),
+        }
+    }
+
+    #[inline]
+    fn safe_slice_ref(&self, start: usize, end: usize) -> Self {
+        Self {
+            ptr: unsafe { self.ptr.add(start) },
+            len: end - start,
+            bytes: self.bytes.clone(),
+        }
+    }
+
+    #[inline]
+    fn as_slice(&self) -> &[u8] {
+        unsafe { std::slice::from_raw_parts(self.ptr, self.len) }
+    }
+}
+
+// Implementations of `UnderlyingBytes` for common types.
+impl UnderlyingBytes for Vec<u8> {}
+impl UnderlyingBytes for Box<[u8]> {}
+impl UnderlyingBytes for String {}
+
+// Implementations of common traits for `Bytes`.
+impl Default for Bytes {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl<T: UnderlyingBytes> From<T> for Bytes {
+    fn from(value: T) -> Self {
+        Self::from_underlying(value)
+    }
+}
+
+impl AsRef<[u8]> for Bytes {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.as_slice()
+    }
+}
+
+impl borrow::Borrow<[u8]> for Bytes {
+    #[inline]
+    fn borrow(&self) -> &[u8] {
+        self.as_slice()
+    }
+}
+
+impl ops::Deref for Bytes {
+    type Target = [u8];
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl<T: AsRef<[u8]>> PartialEq<T> for Bytes {
+    #[inline]
+    fn eq(&self, other: &T) -> bool {
+        self.as_slice() == other.as_ref()
+    }
+}
+
+impl Eq for Bytes {}
+
+impl<T: AsRef<[u8]>> PartialOrd<T> for Bytes {
+    fn partial_cmp(&self, other: &T) -> Option<cmp::Ordering> {
+        self.as_slice().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for Bytes {
+    fn cmp(&self, other: &Bytes) -> cmp::Ordering {
+        self.as_slice().cmp(other.as_slice())
+    }
+}
+
+impl hash::Hash for Bytes {
+    // TODO should we cache the hash since we know the bytes are immutable?
```

Comment:
I suppose we can do that. I suppose the overhead in size is minimal a couple bytes, which are likely anyway within padding (the struct is 32 bytes, and system allocator will give you 16 byte aligned memory...)